### PR TITLE
Slit Controller

### LINKED ISSUE #95: MSA slitlet shapes displayed on cut-out images
Small question of how this is implemented and then another larger question of whether we can use https://github.com/astropy/regions as part of this functionality.

### PR DESCRIPTION
**Pull request by robelgeda**
Closes #95 
This update will allow slits to be drawn using the `regions` package.  It  introduces a way of storing known slit shapes in a [YAML file](https://github.com/spacetelescope/mosviz/blob/d6428c26d7e87705e6d0906570f00e82aa9979d3/mosviz/controls/saved_slits.yaml) as discussed in #95. It also introduces infrastructure for a GUI which the user can use to edit the slit dimensions (added as one of the image viewer tools). More features will be added to the slit editor (apply to all option, slit color, save slit etc...) in the future.

screen shot 2018-08-07 at 2 17 57 am

### Slit selection UI:

screen shot 2018-08-07 at 2 05 02 am

screen shot 2018-08-07 at 2 04 49 am

screen shot 2018-08-07 at 2 05 19 am

screen shot 2018-08-07 at 2 17 46 am

### CODE REVIEW

**Comment by eteq** [comment]
Looks promising!  It me take me a bit to review this as I'll want to try it "live" which means I have to get mosviz back up and running... but hopefully that won't be *too* much work...

**Comment by robelgeda** [comment]
Sounds good to me @eteq! Slack me if you run into bugs. Also, this PR is needed to make MOSViz fully capable again so it might be worth using my branch.

**Comment by robelgeda** [comment]
Ready for review and merger.

File: mosviz/controls/saved_slits.yaml [hunk 1]
```diff
@@ -0,0 +1,44 @@
+nirspec_msa:
+  name: "NIRSpec MSA (Single Slit)"
+  length: 3.3
+  width: 0.2
```
Comment (brechmos-stsci) [inline mosviz/controls/saved_slits.yaml]: @robelgeda Silly question and we might have talked about this, but are units relevant here?  Should unit information be stored so it is super clear what the units are or is it obvious enough to the user?

File: mosviz/controls/slit_controller.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+        if self.is_active:
+            return self.patch.get_width()
+
+    @property
+    def length(self):
+        if self.is_active:
+            return self.patch.get_height()
+
+    @property
+    def dx(self):
```
Comment (brechmos-stsci) [inline mosviz/controls/slit_controller.py]: For the dx and dy, is there a reason to not use property names of `width` and `length` just so it is super clear what they refer to?

File: mosviz/controls/slit_controller.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+        """y axis max and min pixel values"""
+        yp = self.y
+        dy = self.dy
+        y_min = yp - dy / 2.
+        y_max = yp + dy / 2.
+        return [y_min, y_max]
+
+    def construct_simple_rectangle(self, x=None, y=None, width=None, length=None):
+        """
+        Simple matplotlib rectangle patch.
```
Comment (brechmos-stsci) [inline mosviz/controls/slit_controller.py]: For these longer comment blocks we should use the astropy convention of:

```
Parameters
------------

    x:  Pixels
        center of slit in x-axis

    Returns
    --------
        patch:  <type>
            Patch describing the.... 
...
```

File: mosviz/controls/slit_selection_ui.py [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+        name_list = [self._mosviz_table_option_text] + \
+                    [self.slit_dict[s]['name'] for s in sorted(self.slit_dict)] + \
+                    ['Custom']
+
+        key_list = ['default'] + [s for s in sorted(self.slit_dict)] + ['custom']
+
+        combo_input = [(name, key) for name, key in zip(name_list, key_list)]
+        update_combobox(self.slit_type_combo, combo_input, default_index=default_index)
+
+    def get_width(self):
```
Comment (brechmos-stsci) [inline mosviz/controls/slit_selection_ui.py]: Unless this is overloading a method, you should convert this to a property `width`.  The Python convention is to aim for getters that are properties rather than using `get_<variable>` type of thing.

File: mosviz/controls/saved_slits.yaml (hunk 1 above)
Comment (robelgeda) [reply mosviz/controls/saved_slits.yaml]: This is a good idea I will add that

File: mosviz/controls/tests/test_slit_controller.py [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+    :param x: correct center x pixel
+    :param y: correct center y pixel
+    :param width: correct width
+    :param length: correct length
+    """
+
+    x_bounds = np.array([x - (width / 2.), x + (width / 2.)])
+    y_bounds = np.array([y - (length / 2.), y + (length / 2.)])
+
+    assert np.isclose(slit_controller.x, x)
```
Comment (brechmos-stsci) [inline mosviz/controls/tests/test_slit_controller.py]: assert are good for testing and should not be used to do standard checking. You should convert this to an if statement and then raise an Exception explicitly if any of these "fail".  There are some python optimization parameters that will (or used to) remove asserts from code so in those cases these checks will never happen.  I believe it would be best to leave asserts for the test infrastructure and use if/raise type of thing here.

File: mosviz/controls/slit_controller.py (hunk 2 above)
Comment (robelgeda) [reply mosviz/controls/slit_controller.py]: The legacy code uses `dx` and `dy` to refer to the axis in the plots. If we ever decide to use rotated slits the `width` will not match the `dx` of that slit.

Comment (brechmos-stsci) [reply mosviz/controls/slit_controller.py]: Ok, cool, that is what I thought...

File: mosviz/controls/saved_slits.yaml (hunk 1 above)
Comment (eteq) [reply mosviz/controls/saved_slits.yaml]: To make this a bit clearer, what about something like:
```
  length: 3.3 arcsec
  width: 0.2 arcsec
```
instead of what's now here...?  You'll have to pass the values through the `Quantity` initializer, but ``Quantity("3.3 arcsec")`` works out of the box so this should be pretty easy for users to understand...

File: mosviz/controls/slit_controller.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+            width and length of slit in pix.
+        Returns
+        -------
+        patch : `matplotlib.patches.Rectangle`
+        """
+        self.destruct()
+
+        self._slit = None
+        self._pix_slit = None
+        self._patch = Rectangle((x - width / 2, y - length / 2),
```
Comment (eteq) [inline mosviz/controls/slit_controller.py]: Why not instead use the built-in `regions` machinery for doing this? (See https://astropy-regions.readthedocs.io/en/latest/plotting.html , although it looks to me like you're using it further down so maybe you know it already).  That way you don't have to do something special for a slit that's not a definite rectangle.

File: mosviz/controls/slit_controller.py [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+                                width=width, height=length,
+                                edgecolor='red', facecolor='none')
+        return self._patch
+
+    def construct_pix_region(self, x, y, width, length):
```
Comment (eteq) [inline mosviz/controls/slit_controller.py]: I'm confused about why `construct_pix_region` and `construct_sky_region` exist as separate functions that only set the `_pix_slit`/`_slit` and not the other.

One of the core ideas behind regions is that, armed with a WCS, you can convert back and forth between pixel and sky regions (https://astropy-regions.readthedocs.io/en/latest/shapes.html#transformations) - so they shouldn't be "independent" representations - giving one + a WCS you should be able to produce both representations and then only plot the sky one when in sky coordinates and the other in pixel coordinates?

File: mosviz/controls/slit_controller.py [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+from astropy.wcs.utils import proj_plane_pixel_area
+from astropy import units as u
+
+from ..controls.slit_selection_ui import SlitSelectionUI
+
+
+class SlitController(HubListener):
+    """
+    Controller that constructs and stores
+    a rectangular slit.
```
Comment (eteq) [inline mosviz/controls/slit_controller.py]: I realize the name "slit" implies rectangle, but I don't think it's a good idea to restrict this to rectangles.  Part of the point of using regions is that that allows *any* valid region shape to be used here.

It may make sense to initially only support rectangles since that's the most common slit type, but I think you shouldn't limit it to this. In particular, it's important to know that rectangles on the *sky* are not the same as rectangles in pixel space (they are usually distorted).  Also there exist spectrographs (most notably, COS), where the apertures are circular instead of slits.

More broadly, this means you do *not* want to implement the properties like `length`/`width` here, because they are specific to the rectangle region and would give errors if, say, I used a rectangle.

**Comment by robelgeda** [comment]
@eteq : I just saw this! I will go through this slowly and update the code. Thank you so much!!!

**Comment by robelgeda** [comment]
@eteq : I am going to bring this up at today's meeting. There are a couple of open questions here.

File: mosviz/controls/slit_controller.py (hunk 6 above)
Comment (robelgeda) [reply mosviz/controls/slit_controller.py]: This was to provide an API just incase it was needed but I removed it since it is redundant.

File: mosviz/controls/slit_controller.py (hunk 7 above)
Comment (robelgeda) [reply mosviz/controls/slit_controller.py]: I changed the definition of the slit controller and slit:
slit is now a subclass of the `_MOSVizSlit` class. It must implements all abstract attributes and functions.
Slit controller basically has a list of  `_MOSVizSlit` items and is responsible for communicating with the rest of the codebase.

File: mosviz/controls/slit_controller.py (hunk 8 above)
Comment (robelgeda) [reply mosviz/controls/slit_controller.py]: I made the code more generalized for this part.

**Comment by robelgeda** [comment]
@eteq @brechmos-stsci : I updated the slit controls to be more general. Note that slit editor just adds new slits for now since we have **big plans** for it. I will upgrade its functionality in the future.

File: .travis.yml [hunk 9]
```diff
@@ -42,7 +42,7 @@ env:
 
         # List other runtime dependencies for the package that are available as
         # pip packages here.
-        - PIP_DEPENDENCIES='sphinx_rtd_theme sphinx_automodapi specutils==0.2.2 qtpy'
+        - PIP_DEPENDENCIES='sphinx_rtd_theme sphinx_automodapi specutils==0.2.2 qtpy git+https://github.com/astropy/regions.git'
```
Comment (eteq) [inline .travis.yml]: Can you maybe leave a comment here that says *why* this is the dev version of regions?  I.e. "need the work in master as of Sep 11, 2018 which has not yet been released".

Otherwise I fear we will forget why it's a dev version and constantly be saying "hmm, not sure it's safe to switch to a release"...

File: mosviz/controls/slits/rectangle_slit.py [hunk 10]
```diff
@@ -0,0 +1,162 @@
+import numpy as np
+
+from regions import RectangleSkyRegion, RectanglePixelRegion, PixCoord
+from astropy.coordinates import Angle, SkyCoord
+from astropy import units as u
+
+from .slit_base import _MOSVizSlit
+
+
+class _RectangleSlitBase(_MOSVizSlit):
```
Comment (eteq) [inline mosviz/controls/slits/rectangle_slit.py]: Does this really need to be a `Rectangle`?  I might be missing something but it looks general enough that it could be `_RegionsSlitBase` or similar, couldn't it?  Since the drawing is now fully deferred to `regions`?

File: mosviz/controls/slits/saved_slits.yaml [hunk 11]
```diff
@@ -0,0 +1,60 @@
+# Last updated: 09/10/2018
+# name_of_slit:
+#   name: "Display name of slit" # Required
+#   type: MOSViz slit class
+#   length: Quantity or list # User input if not specified
+#   width: Quantity or list # User input if not specified
+
+nirspec_msa:
+  name: "NIRSpec MSA (Single Slit)"
+  type: "RectangleSkySlit"
```
Comment (eteq) [inline mosviz/controls/slits/saved_slits.yaml]: This might be better deferred to another PR if you think it'll take a lot of work... But what about having the type just be "rectangle", and using the type of units in `length` and `width` to determine if it's Sky or Pixel?  (i.e., if it's angular units it's sky, and if it's unitless assume pixel?)

File: mosviz/controls/slits/rectangle_slit.py (hunk 10 above)
Comment (robelgeda) [reply mosviz/controls/slits/rectangle_slit.py]: So for this part it has to be specific for rectangles. This is beacuse some of the functions are specific for rectangle patches (like `self.patch.get_height()` vs `self.patch.get_radius()`).

File: mosviz/controls/slits/saved_slits.yaml (hunk 11 above)
Comment (robelgeda) [reply mosviz/controls/slits/saved_slits.yaml]: I like this, I am going to edit it now.

**Comment by robelgeda** [comment]
Also I accidentally closed the PR 😅

File: mosviz/controls/slits/rectangle_slit.py (hunk 10 above)
Comment (eteq) [reply mosviz/controls/slits/rectangle_slit.py]: Oh, gotcha.  That makes sense... I think some of the things implemented in this class could go up to `_MOSVizSlit`, then (e.g. `patch` and `has_patch`), since they are generic to any region type, not just rectangles.

*But* there's no driving need to do that right now since we are currently only supporting rectangles.  In a future PR allowing other regions this change could be made.
